### DISCUSSION TITLE
将标定参数生成表的一些疑问。

### DISCUSSION THREAD

**Discussion by HaiMianBBao** [Open3DV/Xema#12] [General]
在将标定参数写入到相机里面的时候，会先生成如下的几个参数：
```
xL_rotate_x = xL_undistort_map_x.clone();
xL_rotate_y = xL_undistort_map_y.clone();
rectify_R1 = R1.clone();
pattern_mapping = mapping.clone();
pattern_minimapping = mini_mapping.clone();
```
```
xL_rotate_x, xL_rotate_y, rectify_R1, pattern_mapping, pattern_minimapping
```
这几个参数的作用是什么呢？

相机里面建立点云的时候，用基线除以视差得到z，这个视差是怎么算出来的。

**Comment by vic-w** [comment]
基于DLP结构光的3D成像，在数学上也是使用对极几何，和双目三维重建的数学模型是一样的。在双目三维重建中，我们一般需要左右相机的一对(x,y)坐标，能够计算出三维的（X,Y,Z)坐标。

但DLP结构光与双目重建的不同点在于，实际投射的结构光，一般来说只有竖条纹。我们可以通过对竖条纹解相位，得到投影仪镜头的x坐标，但是投影仪镜头的y坐标我们是无法获得的。实际上，投影仪镜头的y坐标并不是必须的。

设想：我们一般做双目视觉的时候，会用先用畸变矫正和极线校正，将左右图像的极线调整成平行线。完成左右图像的极线校正后，匹配的同名点在左右图像上的Y轴坐标是相同的。我们还需要左右图像的X坐标，就可以算出一个空间点的(X,Y,Z)坐标。也就是说，计算一个空间点的(X,Y,Z)坐标，只需要提供三个数即可，x1，x2和y。这三个数都是指极线校正后的图像中的坐标。

接下来，我们退回到极线校正和畸变矫正之前，此时我们只知道包含畸变的投影仪y坐标和包含畸变的相机的x，y坐标，也知道两个镜头的内外参数。

```
//此时做一下说明：
//在程序里，投影仪的原始带畸变的x坐标记为Xp，p代表projector；相机原始带畸变的坐标记为Xc和Yc，c代表camera
//极限校正后，投影仪的坐标记为Xpr和Ypr，相机的坐标记为Xcr和Ycr，r代表rectify。
```

对于相机这边，我们可以用两个查找表直接把带畸变的Xc,Yc，转换成不带畸变且已经极线校正的Xcr,Ycr。
两个查找表都是二维的查找表，一个输出Xcr，一个输出Ycr。程序中，xL_rotate_x和xL_rotate_y就是这两个查找表。

但到了投影仪这边，就没法这么做了。因为我们缺少了Yp。
盘点一下，我们现在已知Xp，不知道Yp，已知Ypr（因为Ypr=Ycr），而要求的数是Xpr。
我们已知了两个数，要求出一个数，这也可以形成一个二维的查找表。
这个查找表可以根据我们所知的镜头内外参数来建立。程序中，mini_bilinear_interpolation函数就是这个查找表。

此时我们就在不知道Yp的情况下，算出了必要的三个数字Xpr,Xcr,Ycr，可以求出三维的（X,Y,Z）坐标了。

**Comment by Raymonddhr** [comment]
你好，有关相机和投影仪的矫正原理可以讲述一下吗，或者参考的论文可以提供下吗

**Reply by vic-w** [reply]
首先要充分理解双目对极线约束的原理和opencv里面stereoRectify这个函数的功能。
我在网上搜了一个关于极线校正的教程，可以参考
[A Comprehensive Tutorial on Stereo Geometry and Stereo Rectification with Python](https://towardsdatascience.com/a-comprehensive-tutorial-on-stereo-geometry-and-stereo-rectification-with-python-7f368b09924a)
